File: ABSample-ObjC/ABSample-ObjC/AdMediation/AdMediation.m
//
//  AdMediation.m
//  ABSample-ObjC
//
//  Created by Stas Kochkin on 21/06/2018.
//  Copyright © 2018 Appodeal. All rights reserved.
//

#import "AdMediation.h"
#import "AppodealProvider.h"
#import "MoPubProvider.h"


@import Firebase;

static NSString * const kAdMediationTypeKey = @"mediation_partner";

@interface AdMediation ()

@property (nonatomic, assign, readwrite) AdMediationType type;
@property (nonatomic, strong, readwrite) id <AdProvider> provider;
@property (nonatomic, strong) FIRRemoteConfig * config;

@end

@implementation AdMediation

#pragma mark - Designited initializer

- (instancetype)init {
    self = [super init];
    if (self) {
        // Configure Firebase
        [FIRApp configure];
        self.type = AdMediationTypeUnknown;
    }
    return self;
}

+ (instancetype)sharedInstance {
    static AdMediation * _sharedInstance;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _sharedInstance = [AdMediation new];
    });
    return _sharedInstance;
}

#pragma mark - Accessing

- (id<AdProvider>)provider {
    if (!_provider) {
        switch (self.type) {
            case AdMediationTypeAppodeal:
                _provider = [AppodealProvider new];
                break;
            case AdMediationTypeMoPub:
                _provider = [MoPubProvider new];
            default: break;
        }
    }
    return _provider;
}

#pragma mark - Public

- (void)updateMediationType:(void (^)(AdMediationType))completion {
    self.config = FIRRemoteConfig.remoteConfig;
    
    __weak typeof(self) weakSelf = self;
    [self.config fetchWithCompletionHandler:^(FIRRemoteConfigFetchStatus status, NSError * error) {
        AdMediationType type = AdMediationTypeUnknown;
        if (status == FIRRemoteConfigFetchStatusSuccess) {
            [weakSelf.config activateFetched];
            FIRRemoteConfigValue * value = [weakSelf.config configValueForKey:kAdMediationTypeKey];
            if ([value.stringValue isEqualToString:@"appodeal"]) {
                type = AdMediationTypeAppodeal;
            } else if ([value.stringValue isEqualToString:@"mopub"]){
                type = AdMediationTypeMoPub;
            }
        }
        
        weakSelf.type = type;
        completion ? completion(self.type) : nil;
    }];
}


@end

File: ABSample-Swift/ABSample-Swift/AdMediation/AdMediation.swift
//
//  AdMediation.swift
//  ABSample-Swift
//
//  Created by Stas Kochkin on 21/06/2018.
//  Copyright © 2018 Appodeal. All rights reserved.
//

import UIKit
import Firebase

enum AdMediationType : String {
    case Unknown = ""
    case Appodeal = "appodeal"
    case MoPub = "mopub"
}

final class AdMediation: NSObject {
    private struct Constants {
        static let key = "mediation_partner"
    }
    
    private(set) var type: AdMediationType = .Unknown
    private lazy var config : RemoteConfig = {
       return RemoteConfig.remoteConfig()
    }()
    
    private(set) var provider : AdProvider?
    
    static let sharedInstance : AdMediation = AdMediation()
    
    override init() {
        FirebaseApp.configure()
    }
    
    func update(_ completion : @escaping (AdMediationType)->()) {
        config.fetch { [unowned self] status, error in
            guard  status == RemoteConfigFetchStatus.success else {
                print("Error while fetching remote config \(error.map{$0.localizedDescription} ?? "unknown")")
                completion(self.type)
                return
            }
            self.config.activateFetched()
            let value = self.config.configValue(forKey: Constants.key)
            self.type = value.stringValue.map { AdMediationType(rawValue: $0)}! ?? .Unknown
            // Update provider
            switch self.type {
            case .Appodeal:
                self.provider = AppodealProvider()
            case .MoPub:
                self.provider = MoPubProvider()
            case .Unknown:
                break
            }
            
            completion(self.type)
        }
    }
}
